src/core/net/tcp6.hpp

```diff
@@ -0,0 +1,302 @@
+
+        /**
+         * This method returns the TCP Source Port.
+         *
+         * @returns The TCP Source Port.
+         *
+         */
+        uint16_t GetSourcePort(void) const { return HostSwap16(mSource); }
+
+        /**
+         * This method returns the TCP Destination Port.
+         *
+         * @returns The TCP Destination Port.
+         *
+         */
+        uint16_t GetDestinationPort(void) const { return HostSwap16(mDestination); }
+
+        /**
+         * This method returns the TCP Sequence Number.
+         *
+         * @returns The TCP Sequence Number.
+         *
+         */
+        uint32_t GetSequenceNumber(void) const { return HostSwap32(mSequenceNumber); }
+
+        /**
+         * This method returns the TCP Acknowledgment Sequence Number.
+         *
+         * @returns The TCP Acknowledgment Sequence Number.
+         *
+         */
+        uint32_t GetAcknowledgmentNumber(void) const { return HostSwap32(mAckNumber); }
+
+        /**
+         * This method returns the TCP Flags.
+         *
+         * @returns The TCP Flags.
+         *
+         */
+        uint16_t GetFlags(void) const { return HostSwap16(mFlags); }
+
+        /**
+         * This method returns the TCP Window.
+         *
+         * @returns The TCP Window.
+         *
+         */
+        uint16_t GetWindow(void) const { return HostSwap16(mWindow); }
+
+        /**
+         * This method returns the TCP Checksum.
+         *
+         * @returns The TCP Checksum.
+         *
+         */
+        uint16_t GetChecksum(void) const { return HostSwap16(mChecksum); }
+
+        /**
+         * This method returns the TCP Urgent Pointer.
+         *
+         * @returns The TCP Urgent Pointer.
+         *
+         */
+        uint16_t GetUrgentPointer(void) const { return HostSwap16(mUrgentPointer); }
+
+    private:
+        uint16_t mSource;
+        uint16_t mDestination;
+        uint32_t mSequenceNumber;
+        uint32_t mAckNumber;
+        uint16_t mFlags;
+        uint16_t mWindow;
+        uint16_t mChecksum;
+        uint16_t mUrgentPointer;
+
+    } OT_TOOL_PACKED_END;
+
+    /**
+     * This constructor initializes the object.
+     *
+     * @param[in] aInstance  A reference to the OpenThread instance.
+     *
+     */
+    explicit Tcp(Instance &aInstance);
+
+    /**
+     * See documentation for otTcpEndpointInitialize in include/openthread/tcp.h.
+     */
+    Error Initialize(Endpoint &aEndpoint, otTcpEndpointInitializeArgs &aArgs);
+
+    /**
+     * See documentation for otTcpEndpointGetInstance in include/openthread/tcp.h.
+     */
+    static Instance *GetInstance(Endpoint &aEndpoint);
+
+    /**
+     * See documentation for otTcpEndpointGetContext in include/openthread/tcp.h.
+     */
+    static void *GetContext(Endpoint &aEndpoint);
+
+    /**
+     * See documentation for otTcpGetLocalAddress in include/openthread/tcp.h.
+     */
+    static const SockAddr &GetLocalAddress(Endpoint &aEndpoint);
```

Comment:
How about we change these to be methods of the `Endpoint` class?

Reply:
I agree that moving these member functions into the `Endpoint` class is natural, and I'm happy to make this change.

The reason I had originally put them into the `Tcp` class with `Endpoint &` as the first argument is that `Udp` did it this way. If you look at `src/core/net/udp6.cpp`, the functions are implemented in the `Udp` class, with methods in the `Listener` and `SocketHandle` classes calling the functions in the `Udp` class.

Reply:
FYI, this would change the flow a little bit, as some of these functions will need to access members of the `Tcp` class, so looking up the instance would move from `tcp_api.cpp` here.

Reply:
I think having them defined in `Endpoint`/`Listener` class would be simpler 

Since we already assume/require that `Endpoint` and `Listener` track the `ot::Instance` they belong to, they can then get to any other object in OT from `Instance` itself (using the `Instance::Get<>()` specialization for different types, e.g. `Get<Tcp>()` to get the `Tcp` class). 

I won't recommend the `Udp` socket model as a good example to follow. The model we have used there is more tied to some issues in the first API design and implementation and us wanting to keep the public APIs unchanged later on. You can see more on this from Issue https://github.com/openthread/openthread/issues/5162 and this PR https://github.com/openthread/openthread/pull/5231.

One thing I suggest to avoid the same issue that came up in the `Udp::Socket` case is that we don't declare `Endpoint` and `Listener` as `InstanceLocator` and instead have `otInstance *mInstance` as a member variable directly included in the `/* Other implementation-defined fields go here. */` section of the original `struct otTcpEndpoint` or `struct otTcpLisner`.